### PR TITLE
Add RemoteService to FeignContext

### PR DESCRIPTION
**Pull request by wapkch**
As mentioned in #2040, i think RemoteService should be set to FeignContext, so components like Skywalking could retrieve the host and port as the peer instead of using templated URL path.

@marcingrzejszczak Please review this draft PR.

### CODE REVIEW

**Review by velo — APPROVED** [review]

**Comment by marcingrzejszczak** [comment]
Hey! How does it look like in Zipkin after your changes? Do you have a screenshot?

**Comment by wapkch** [comment]
> Hey! How does it look like in Zipkin after your changes? Do you have a screenshot?

Sure, before the changes:
711685610291_ pic

After the changes:
701685610115_ pic

The FeignClient:
```java
@FeignClient(name = "jsonplaceholder", url = "jsonplaceholder.typicode.com")
public interface JsonPlaceholderFeignClient {

    @GetMapping("/posts")
    List<Post> loadPosts();

}

```

**Comment by wapkch** [comment]
> What will happen if the other application (e.g. `jsonplaceholder`) is also instrumented and has a different name (e.g. `foo`)? The standard practice is that the service name gets automatically resolved as the server's local service name. Can you test this against an app (e.g. a Spring Boot 3 app) that has an MVC filter that is instrumented and show the screenshot?

It will be like this:
image
In this test, `jsonplaceholder` application's `spring.application.name` is `foo`, and the FeignClient's name is still `jsonplaceholder`

**Comment by marcingrzejszczak** [comment]
So this is with the changes from this PR? Cause the second span still has `application` there as a service name?

**Comment by wapkch** [comment]
> So this is with the changes from this PR? Cause the second span still has `application` there as a service name?

Yes, i confirmed, the second span just had a different service name compared with the case that the remote service did not report traces.

**Comment by marcingrzejszczak** [comment]
I'm asking cause this works well if you're not using apache skywalking. WDYT about doing the setting of remote service name and address inside skywalking via the observation convention mechanism? Or the observation filter?

**Comment by wapkch** [comment]
> I'm asking cause this works well if you're not using apache skywalking. WDYT about doing the setting of remote service name and address inside skywalking via the observation convention mechanism? Or the observation filter?

I considered this option, but i think the problem of doing the setting inside skywalking is that skywalking have to distinguish the different Sender implementations(Retrieve the remote service name and address from carrier then set to SenderContext), there are so many SenderContext and carrier implementations, it will be hard to maintain in skywalking. So i think doing the settings in seperated projects will be better.

**Comment by marcingrzejszczak** [comment]
>  there are so many SenderContext and carrier implementations(e.g., Feign, Resttemplate, messaging...)

So you're saying that there will be the same problem with all of these implementations? I'm just curious if this is not a problem related to Skywalking as such. I don't recall having such problems with Zipkin / Tempo etc.

File: micrometer/src/main/java/feign/micrometer/FeignContext.java [hunk 1]
```diff
@@ -29,6 +31,15 @@ public class FeignContext extends RequestReplySenderContext<Request, Response> {
   public FeignContext(Request request) {
     super(Request::header);
     setCarrier(request);
+    setRemoteServiceName(request.requestTemplate().feignTarget().name());
+    try {
+      URL url = new URL(request.requestTemplate().feignTarget().url());
+      if (StringUtils.isNotBlank(url.getProtocol()) && StringUtils.isNotBlank(url.getHost())) {
+        setRemoteServiceAddress(url.getProtocol() + "://" + url.getHost() + ":" + url.getPort());
```
Comment (velo) [inline micrometer/src/main/java/feign/micrometer/FeignContext.java:38]: Wouldn't `setRemoteServiceAddress(url.toString())` do the trick?

File: micrometer/src/main/java/feign/micrometer/FeignContext.java [hunk 2]
```diff
@@ -29,6 +31,15 @@ public class FeignContext extends RequestReplySenderContext<Request, Response> {
...
+      }
+    } catch (Exception ignored) {
```
Comment (velo) [inline micrometer/src/main/java/feign/micrometer/FeignContext.java:40]: what exception is been ignored here?

probably should re-throw it, as will be impossible for people to understand why something is not working if we are eating the exception

**Review by velo — CHANGES_REQUESTED** [review]
When ready, please move from draft

**Comment by velo** [comment]
No answer to feedback and still on draft, will close, if needed please reopen

**Pull request closed without merging**